omniledger/darc/request.go

```diff
@@ -0,0 +1,107 @@
+package darc
+
+import (
+	"crypto/sha256"
+	"errors"
+)
+
+// Hash computes the digest of the request, the identities and signatures are
+// not included.
+func (r *Request) Hash() []byte {
+	return r.innerRequest.Hash()
+}
+
+func (r innerRequest) Hash() []byte {
+	h := sha256.New()
+	h.Write(r.BaseID)
+	h.Write([]byte(r.Action))
+	h.Write(r.Msg)
+	for _, i := range r.Identities {
+		h.Write([]byte(i.String()))
+	}
+	return h.Sum(nil)
+}
+
+// GetIdentityStrings returns a slice of identity strings, this is useful for
+// creating a parser.
+func (r *Request) GetIdentityStrings() []string {
+	res := make([]string, len(r.Identities))
+	for i, id := range r.Identities {
+		res[i] = id.String()
+	}
+	return res
+}
+
+// MsgToDarc attempts to return a darc given the matching darcBuf. This
+// function should *not* be used as a way to verify the darc, it only checks
+// that darcBuf can be decoded and matches with the Msg part of the request.
+func (r *Request) MsgToDarc(darcBuf []byte) (*Darc, error) {
+	d, err := NewDarcFromProto(darcBuf)
+	if err != nil {
+		return nil, err
+	}
+
+	if !d.GetID().Equal(r.Msg) {
+		return nil, errors.New("darc IDs are not equal")
+	}
+
+	if len(r.Signatures) != len(r.Identities) {
+		return nil, errors.New("signature and identitity length mismatch")
+	}
+	darcSigs := make([]*Signature, len(r.Signatures))
+	for i := range r.Signatures {
+		darcSigs[i] = &Signature{
+			Signature: r.Signatures[i],
+			Signer:    *r.Identities[i],
+		}
+	}
+	d.Signatures = darcSigs
+
+	return d, nil
+}
+
+// InitRequest initialises a request, the caller must provide all the fields of
+// the request. There is no guarantee that this request is valid, please see
+// InitAndSignRequest is a valid request needs to be created.
```

Comment:
s/is/if/

Reply:
Maybe these two should be called "InitRequestWithSignatures" and "InitRequestAndSign" in order to explain the difference better. Also a sentence in the docs explaining why you'd use one vs the other would be helpful. (I can't propose it because I can't understand it yet.)

Reply:
Why not `NewRequest` and `NewRequestAndSign`?